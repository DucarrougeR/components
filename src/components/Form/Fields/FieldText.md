```js noeditor
const StatusAndResources = require('../../../../styleguide_components/StatusAndResources').StatusAndResources;

<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/2MG6DoSjk4IaLnjjFCnKFf/Forms"
  githubURL="https://github.com/looker/relens/blob/master/src/components/Form/Fields/FieldText.tsx"
  feedbackTitle="FieldText Component Feedback" />
```

```js
  <FieldText id="theme-name" name="themeName" label="Theme Name" />
```

```js

<form>
  <FieldText id="first_name" name="first_name" label="First Name" />
  <FieldText id="last_name" name="last_name" label="Last Name" />
  <FieldCheckbox id="subscribe" name="subscribe" label="Subscribe" />
  <Button>Submit</Button>
</form>
```
