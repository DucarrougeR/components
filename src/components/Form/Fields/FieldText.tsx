import * as React from 'react'
import { FieldJSX, FieldProps } from './Field'

export const FieldText: React.SFC<FieldProps> = ({
  id,
  label,
  name,
  alignLabel,
  validationState,
  alignValidationMessage,
  ...props
}) => {
  return FieldJSX<typeof props>(
    'FieldText',
    id,
    label,
    name,
    alignLabel,
    validationState,
    alignValidationMessage,
    props
  )
}
