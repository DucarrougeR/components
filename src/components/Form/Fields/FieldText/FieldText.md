```js noeditor
const StatusAndResources = require('../../../../../styleguide_components/StatusAndResources').StatusAndResources;

<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/2MG6DoSjk4IaLnjjFCnKFf/Forms"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Form/Fields/FieldText.tsx"
  feedbackTitle="FieldText Component Feedback" />
```

### FieldText

The `<FieldText />` component is composed of an `<InputText />` component and a `<Label />` component. Using `<FieldText />` allows for rendering validation messages. By default, the label will render directly above the input field, which is the recommended convention, however this is adjustable with the `alignLabel` property.

```js
<FieldText name="firstName" label="First Name" />
<FieldText name="lastName" label=" Last Name" alignLabel="left" />
<FieldText name="requiredField" label="A required field" required />
```

### FieldText Validation

The `<FieldText />` component accepts `validationMessage` as a property. This is an object with properties `type` and `message`.

See `ValidationMessage` [here](/#!/ValidationMessage)

If you have multiple fields in need of validation, it is recommended that you wrap them in a `<Form />` component and pass to it a dictionary with key as field name and value as a `ValidationMessage`.

___Note: In the current implementation, any `ValidationMessage` set in a `<FieldText />` component will be overridden by `ValidationMessages` set in the parent `<Form />` component. Because of this, it is highly recommended that you only ever directly set `ValidationMessage` if a `<FieldText />` is outside a `<Form />`.___

```js
<FieldText name="someField" label="Some Field" validationMessage={{type: "error", message: "This is an error"}}/>
<FieldText name="someField2" label="Some Field2" validationMessage={{type: "warning", message: "Here is a warning"}} alignValidationMessage="right" />
```
