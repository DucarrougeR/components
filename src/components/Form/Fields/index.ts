export * from './FieldCheckbox'
export * from './FieldText'
