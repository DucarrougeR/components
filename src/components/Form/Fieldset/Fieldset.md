```js noeditor
const StatusAndResources = require('../../../../styleguide_components/StatusAndResources').StatusAndResources;

<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/2MG6DoSjk4IaLnjjFCnKFf/Forms"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Form/Fieldset/Fieldset.tsx"
  feedbackTitle="Fieldset Component Feedback" />
```

### Fieldset

The `<Fieldset />` component is used to wrap Field components into meaningful groups. The `legend` property allows for a heading which is by default aligned above and to the left.

```js
<Fieldset legend="Legend">
  <FieldText label="First" />
  <FieldText label="Second" />
  <FieldText label="Third" />
</Fieldset>
<Fieldset legend="Legend2" alignLabels='left' labelWidth='40%'>
  <FieldText label="First" />
  <FieldText label="Second" />
  <FieldText label="Third" />
</Fieldset>
<Button>Submit</Button>
```
