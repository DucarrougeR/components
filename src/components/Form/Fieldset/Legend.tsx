import * as React from 'react'
import { SpacingSizes } from '../../../style'
import { fontWeights } from '../../../style/font_weights'
import { CustomizableAttributes } from '../../../types/attributes'
import { Box, BoxProps } from '../../Box'

export const Legend: React.SFC<BoxProps<HTMLLegendElement>> = ({
  ...props
}) => (
  <Box
    is="legend"
    color={CustomizableLegendAttributes.color}
    fontSize={CustomizableLegendAttributes.fontSize}
    fontWeight={CustomizableLegendAttributes.fontWeight}
    pb={CustomizableLegendAttributes.bottomPadding}
    {...props}
  />
)

interface CustomizableLegendAttributes extends CustomizableAttributes {
  bottomPadding: SpacingSizes
}

export const CustomizableLegendAttributes: CustomizableLegendAttributes = {
  bottomPadding: 'xsmall',
  color: 'palette.charcoal800',
  fontSize: 'xxxlarge',
  fontWeight: fontWeights.light,
}
