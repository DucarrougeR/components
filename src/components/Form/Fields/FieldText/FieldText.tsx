import React, { FunctionComponent, Ref } from 'react'
import styled, { StyledComponent } from 'styled-components'
import uuid from 'uuid/v4'
import { ComponentWithForm, withForm } from '../../Form'
import { InputText, InputTextProps } from '../../Inputs/InputText/InputText'
import { Field, FieldProps } from '../Field'

export interface FieldTextProps extends FieldProps, InputTextProps {}
type ComponentType = FunctionComponent<FieldTextProps>
type StyledComponentType = StyledComponent<ComponentType, FieldTextProps>

const InternalFieldText: ComponentType = (props: FieldTextProps) => {
  const {
    alignLabel,
    alignValidationMessage,
    label,
    labelWidth,
    labelFontWeight,
    validationMessage,
    id = uuid(),
    ...inputTextProps
  } = props
  return (
    <Field
      id={id}
      {...props}
      alignValidationMessage={
        alignValidationMessage ? alignValidationMessage : 'bottom'
      }
    >
      <InputText
        id={id}
        {...inputTextProps}
        validationType={validationMessage && validationMessage.type}
      />
    </Field>
  )
}

const FieldTextFactory = React.forwardRef<StyledComponentType, FieldTextProps>(
  (props: FieldTextProps, ref: Ref<StyledComponentType>) => (
    <InternalFieldText ref={ref} {...props} />
  )
)

/** @component */
export const FieldText: StyledComponent<
  ComponentWithForm<FieldTextProps>,
  FieldTextProps
> = styled<ComponentWithForm<FieldTextProps>>(
  withForm<FieldTextProps>(FieldTextFactory)
)``
