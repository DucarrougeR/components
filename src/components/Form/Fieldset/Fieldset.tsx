import * as React from 'react'
import { Box, BoxProps } from '../../Box'
import { FlexItem } from '../../FlexItem'
import { FormControl, FormControlDirections } from '../FormControl'
import { Legend } from './Legend'

export interface FieldsetProps extends BoxProps<HTMLFieldSetElement> {
  /**
   * Specifies where to render the legend in relation to the set of inputs. Can be placed `left`, `right`, `bottom`, or `top`.
   */
  alignLegend?: FormControlDirections
  /**
   * The legend, or heading, of this fieldset.
   */
  legend?: string
}

export const Fieldset: React.FC<FieldsetProps> = ({ legend, ...props }) => {
  return (
    <Box is="fieldset" {...props}>
      <FormControl mb="xsmall" alignLabel={props.alignLegend}>
        {legend ? (
          <FlexItem>
            <Legend>{legend}</Legend>
          </FlexItem>
        ) : null}
        <FlexItem>{props.children}</FlexItem>
      </FormControl>
    </Box>
  )
}
