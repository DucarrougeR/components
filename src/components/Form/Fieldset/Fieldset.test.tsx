import 'jest-styled-components'
import * as React from 'react'
import { createWithTheme } from '../../../../test/utils/create_with_theme'
import { FieldText } from '../Fields/FieldText'
import { Fieldset } from './Fieldset'

test('Fieldset with left aligned legend', () => {
  const component = createWithTheme(
    <Fieldset legend="Legend" alignLegend="left">
      <FieldText label="One" name="name1" />
      <FieldText label="two" name="name2" />
      <FieldText label="three" name="nam3" />
    </Fieldset>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Fieldset with left aligned labels and custom label width', () => {
  const component = createWithTheme(
    <Fieldset legend="Legend" alignLabels="left" labelWidth="30%">
      <FieldText label="One" name="name1" />
      <FieldText label="two" name="name2" />
      <FieldText label="three" name="nam3" />
    </Fieldset>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
