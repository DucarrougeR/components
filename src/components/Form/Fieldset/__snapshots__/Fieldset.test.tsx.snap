// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fieldset with left aligned legend 1`] = `
.c0 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
}

.c1 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  -webkit-animation: none;
  animation: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-bottom: 0.5rem;
}

.c2 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  color: #262D33;
  font-size: 2.25rem;
  font-weight: 300;
  padding-bottom: 0.5rem;
}

.c3 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 0.5rem;
}

.c4 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  color: #262D33;
  font-size: 0.875rem;
  font-weight: 600;
  margin-right: 0.5rem;
}

.c5 {
  border: 0;
  box-sizing: border-box;
  font: inherit;
  font-size: 100%;
  margin: 0;
  padding: 0;
  vertical-align: baseline;
  -webkit-animation: none;
  animation: none;
  border: solid 1px;
  border-radius: 0.25rem;
  border-color: #C1C6CC;
  font-size: 0.875rem;
  height: 28px;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
}

<fieldset
  alignLegend="left"
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c0"
    >
      <legend
        className="c2"
      >
        Legend
      </legend>
    </div>
    <div
      className="c0"
    >
      <div
        className="c3"
      >
        <label
          className="c4"
          htmlFor="text-1"
        >
          One
        </label>
        <div
          className="c3"
        >
          <div
            className="c0"
          >
            <input
              className="c5"
              id="text-1"
              name="name1"
              type="text"
            />
          </div>
        </div>
      </div>
      <div
        className="c3"
      >
        <label
          className="c4"
          htmlFor="text-2"
        >
          two
        </label>
        <div
          className="c3"
        >
          <div
            className="c0"
          >
            <input
              className="c5"
              id="text-2"
              name="name2"
              type="text"
            />
          </div>
        </div>
      </div>
      <div
        className="c3"
      >
        <label
          className="c4"
          htmlFor="text-3"
        >
          three
        </label>
        <div
          className="c3"
        >
          <div
            className="c0"
          >
            <input
              className="c5"
              id="text-3"
              name="nam3"
              type="text"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</fieldset>
`;
